File: state/db/models/notes.py
"""
Notes models with encryption.

Notes are encrypted with the user's passphrase for privacy.
"""

from datetime import datetime

from sqlalchemy import DateTime, Index, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base


class Note(Base):
    """
    Category or group note.

    Content is encrypted with user's passphrase.
    Supports note inheritance (most recent note applies to future months).
    """

    __tablename__ = "notes"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)  # UUID
    category_type: Mapped[str] = mapped_column(String(20), nullable=False)  # 'group' or 'category'
    category_id: Mapped[str] = mapped_column(String(100), nullable=False)  # Monarch ID
    category_name: Mapped[str] = mapped_column(String(255), nullable=False)
    group_id: Mapped[str | None] = mapped_column(String(100), nullable=True)
    group_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    month_key: Mapped[str] = mapped_column(String(10), nullable=False)  # "2025-01"
    # Encrypted content - Fernet ciphertext of markdown
    content_encrypted: Mapped[str] = mapped_column(Text, nullable=False)
    # Salt for this note's encryption (allows re-keying individual notes if needed)
    salt: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    __table_args__ = (
        Index("idx_notes_category", "category_type", "category_id"),
        Index("idx_notes_month", "month_key"),
    )


class GeneralNote(Base):
    """
    General note for a month (not tied to a category).

    Content is encrypted with user's passphrase.
    """

    __tablename__ = "general_notes"

    month_key: Mapped[str] = mapped_column(String(10), primary_key=True)  # "2025-01"
    id: Mapped[str] = mapped_column(String(36), nullable=False)  # UUID
    # Encrypted content
    content_encrypted: Mapped[str] = mapped_column(Text, nullable=False)
    salt: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)


class ArchivedNote(Base):
    """
    Archived note from a deleted category.

    Preserves notes when categories are deleted from Monarch.
    Content remains encrypted.
    """

    __tablename__ = "archived_notes"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)  # UUID
    category_type: Mapped[str] = mapped_column(String(20), nullable=False)
    category_id: Mapped[str] = mapped_column(String(100), nullable=False)
    category_name: Mapped[str] = mapped_column(String(255), nullable=False)
    group_id: Mapped[str | None] = mapped_column(String(100), nullable=True)
    group_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    month_key: Mapped[str] = mapped_column(String(10), nullable=False)
    # Encrypted content
    content_encrypted: Mapped[str] = mapped_column(Text, nullable=False)
    salt: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    archived_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    original_category_name: Mapped[str] = mapped_column(String(255), nullable=False)
    original_group_name: Mapped[str | None] = mapped_column(String(255), nullable=True)


class KnownCategory(Base):
    """
    Tracks known Monarch categories for deletion detection.

    When a category disappears from Monarch, we can detect it
    and archive related notes.
    """

    __tablename__ = "known_categories"

    category_id: Mapped[str] = mapped_column(String(100), primary_key=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)

File: frontend/src/types/notes.ts
/**
 * Monthly Notes Type Definitions
 *
 * Types for the Monthly Notes feature that allows users to attach
 * markdown notes to Monarch category groups and categories.
 * Notes support time-based inheritance and are stored locally.
 */

/**
 * Month key format: "YYYY-MM" (e.g., "2025-01")
 * Used as the primary identifier for month-based lookups.
 *
 * This branded type provides:
 * 1. Semantic documentation that this is specifically a month key, not any string
 * 2. Intentional friction when passing arbitrary strings as month keys
 *
 * Use `as MonthKey` when creating month keys from known valid sources.
 */
export type MonthKey = string & { readonly __brand?: 'MonthKey' };

/**
 * Reference to a category or category group in Monarch
 */
export interface CategoryReference {
  /** Whether this is a group or individual category */
  type: 'group' | 'category';
  /** Monarch category_id or group_id */
  id: string;
  /** Display name */
  name: string;
  /** For categories, their parent group ID */
  groupId?: string;
  /** For categories, their parent group name */
  groupName?: string;
}

/**
 * A single note entry attached to a category or group
 */
export interface Note {
  /** UUID for the note */
  id: string;
  /** Reference to the category or group this note is attached to */
  categoryRef: CategoryReference;
  /** The month when this note was explicitly set (not inherited) */
  monthKey: MonthKey;
  /** Markdown content of the note */
  content: string;
  /** ISO timestamp when note was created */
  createdAt: string;
  /** ISO timestamp when note was last modified */
  updatedAt: string;
}

/**
 * General notes for a month (not tied to any category)
 */
export interface GeneralMonthNote {
  /** UUID for the note */
  id: string;
  /** The month this note is for */
  monthKey: MonthKey;
  /** Markdown content of the note */
  content: string;
  /** ISO timestamp when note was created */
  createdAt: string;
  /** ISO timestamp when note was last modified */
  updatedAt: string;
}

/**
 * Metadata about a month's notes
 */
export interface MonthMetadata {
  /** The month key */
  monthKey: MonthKey;
  /** Most recent edit to any note this month (category or general) */
  lastUpdatedAt: string | null;
}

/**
 * An archived note (category was deleted from Monarch)
 */
export interface ArchivedNote extends Note {
  /** ISO timestamp when note was archived */
  archivedAt: string;
  /** Original category name before deletion */
  originalCategoryName: string;
  /** Original group name before deletion (for categories) */
  originalGroupName?: string;
}

/**
 * The effective note for display, which may be inherited from an earlier month.
 * This is the result of the inheritance lookup algorithm.
 */
export interface EffectiveNote {
  /** The note content, or null if no note exists */
  note: Note | null;
  /** Which month the note originates from */
  sourceMonth: MonthKey | null;
  /** True if note is inherited from an earlier month */
  isInherited: boolean;
}

/**
 * The effective general note for display, which may be inherited from an earlier month.
 * Same concept as EffectiveNote but for general month notes.
 */
export interface EffectiveGeneralNote {
  /** The general note content, or null if no note exists */
  note: GeneralMonthNote | null;
  /** Which month the note originates from */
  sourceMonth: MonthKey | null;
  /** True if note is inherited from an earlier month */
  isInherited: boolean;
}

/**
 * A category group with its effective note and child categories
 */
export interface CategoryGroupWithNotes {
  /** Monarch group ID */
  id: string;
  /** Group display name */
  name: string;
  /** The effective note for this group */
  effectiveNote: EffectiveNote;
  /** Child categories with their notes */
  categories: CategoryWithNotes[];
  /** Whether this group is expanded in the UI */
  isExpanded: boolean;
}

/**
 * A category with its effective note
 */
export interface CategoryWithNotes {
  /** Monarch category ID */
  id: string;
  /** Category display name */
  name: string;
  /** Parent group ID */
  groupId: string;
  /** Category icon/emoji if available */
  icon?: string;
  /** The effective note for this category */
  effectiveNote: EffectiveNote;
}

/**
 * Full state for a month view
 */
export interface MonthNotesState {
  /** The month being viewed */
  monthKey: MonthKey;
  /** Metadata about this month */
  metadata: MonthMetadata;
  /** Category groups with their notes */
  groups: CategoryGroupWithNotes[];
  /** Effective general month note (may be inherited from earlier month) */
  effectiveGeneralNote: EffectiveGeneralNote | null;
  /** Archived notes from deleted categories */
  archivedNotes: ArchivedNote[];
}

/**
 * A note version entry for revision history
 */
export interface NoteVersion {
  /** The month when this version was set */
  monthKey: MonthKey;
  /** Preview of the note content (first ~100 chars) */
  contentPreview: string;
  /** Full content */
  content: string;
  /** Whether this is the currently viewed month */
  isCurrent: boolean;
  /** ISO timestamp when this version was created */
  createdAt: string;
}

/**
 * Revision history for a category or group's notes
 */
export interface NoteRevisionHistory {
  /** Reference to the category/group */
  categoryRef: CategoryReference;
  /** All versions of notes for this category/group, sorted by month */
  versions: NoteVersion[];
}

// ============================================================================
// API Request/Response Types
// ============================================================================

/**
 * Request to save a category note
 */
export interface SaveCategoryNoteRequest {
  /** Type of target (group or category) */
  categoryType: 'group' | 'category';
  /** Monarch ID of the category or group */
  categoryId: string;
  /** Display name */
  categoryName: string;
  /** For categories, parent group ID */
  groupId?: string;
  /** For categories, parent group name */
  groupName?: string;
  /** Month to save the note for */
  monthKey: MonthKey;
  /** Markdown content */
  content: string;
}

/**
 * Request to save a general month note
 */
export interface SaveGeneralNoteRequest {
  /** Month to save the note for */
  monthKey: MonthKey;
  /** Markdown content */
  content: string;
}

/**
 * Request to export notes as PDF
 */
export interface ExportNotesRequest {
  /** Start month (inclusive) */
  startMonth: MonthKey;
  /** End month (inclusive) */
  endMonth: MonthKey;
  /** Whether to include category notes */
  includeCategoryNotes: boolean;
  /** Whether to include general month notes */
  includeGeneralNotes: boolean;
}

/**
 * Response from saving a note
 */
export interface SaveNoteResponse {
  /** Whether the save was successful */
  success: boolean;
  /** The saved note */
  note?: Note | GeneralMonthNote;
  /** Error message if save failed */
  error?: string;
}

/**
 * Response from getting month notes
 */
export interface GetMonthNotesResponse {
  /** The month notes state */
  data: MonthNotesState;
}

/**
 * Response from getting revision history
 */
export interface GetRevisionHistoryResponse {
  /** The revision history */
  history: NoteRevisionHistory;
}

// ============================================================================
// Checkbox State Types
// ============================================================================

/**
 * Checkbox states for a note
 */
export interface CheckboxStates {
  /** Array of checked states, indexed by checkbox position in markdown */
  states: boolean[];
}

/**
 * Request to update a checkbox state
 */
export interface UpdateCheckboxRequest {
  /** Note ID for category/group notes */
  noteId?: string;
  /** Month key for general notes */
  generalNoteMonthKey?: string;
  /** The month being viewed */
  viewingMonth: string;
  /** Checkbox position in the markdown (0-indexed) */
  checkboxIndex: number;
  /** New checked state */
  isChecked: boolean;
}

/**
 * Response from getting checkbox states
 */
export interface GetCheckboxStatesResponse {
  /** Array of checkbox states */
  states: boolean[];
}

/**
 * Response from getting all checkbox states for a month
 */
export interface GetMonthCheckboxStatesResponse {
  /** Map of note_id -> checkbox states */
  states: Record<string, boolean[]>;
}

/**
 * Response from updating checkbox state
 */
export interface UpdateCheckboxResponse {
  success: boolean;
  /** Updated checkbox states */
  states: boolean[];
}

// ============================================================================
// Notes Categories Types
// ============================================================================

/**
 * A category within a group for the notes feature
 */
export interface NotesCategory {
  /** Monarch category ID */
  id: string;
  /** Category display name */
  name: string;
  /** Category icon/emoji if available */
  icon?: string;
}

/**
 * A category group with its categories for the notes feature
 */
export interface NotesCategoryGroup {
  /** Monarch group ID */
  id: string;
  /** Group display name */
  name: string;
  /** Categories within this group */
  categories: NotesCategory[];
}

/**
 * Response from getting all notes categories
 */
export interface GetNotesCategoriesResponse {
  /** All category groups with their categories */
  groups: NotesCategoryGroup[];
}
